### PR TITLE
new: introduce `is_exe_upper_layer`

### PR DESCRIPTION
**Pull request by loresuso**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area driver-kmod

/area driver-ebpf

/area libscap

/area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:

This PR tries to deal with `overlayfs`, a union mount filesystem often used by container runtimes, to understand if a process is executing something that is part of the base image of a container or not. Monitoring this could be relevant from a security standpoint since best practices state that containers should be immutable. For this reason, in many use cases, we don't expect processes in the container to execute something that is not part of the base image, whether that file was created in the container or an existing one was modified. 
To do so, the PR aims at identifying the layer of the file being executed by a process: if the file belongs to the lower layer, it means that it was part of the base image of the container. Otherwise, the process executed something that is part of the upper layer, and a flag is set upon the `execve` event.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

To better understand the code, I'd like to give reviewers some useful pointers. First, we must check if the file being executed is part of an overlayfs. We can do that by retrieving the superblock magic of the file and comparing it with the `OVERLAYFS_SUPER_MAGIC`.  You can see [here](https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/super.c#L2129) that this constant is used when an overlayfs is mounted. 
Then, to check the layer of the executable file, we can do some pointer arithmetics, having in mind that the pointer to the inode that we can retrieve form `struct file` is pointing to the `vfs_inode` member of the `ovl_inode` struct, as you can see [here](https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/ovl_entry.h#L122).

The last thing that I wanted to discuss is that for now I didn't find a way to retrieve the upper_layer information when scanning the proc filesystem. One possible way to do it could be by inspecting the `/proc/<pid>/mounts` file and checking if there is an overlayfs entry, which in case we can use to retrieve the upperdir directory and look for the `exe` file there. However, this would easily work when Falco is installed on the host, but could be a problem when it is running as a container. This is because most of the time the upperdir directory is placed in `/var/lib/docker`, `/var/lib/containerd` directories. This means we would need to give the Falco container access to those directories. I would like to know what do you think, and if you like it, I can implement it. Also, I am open to any other possible solution on this!

EDIT: I have implemented this approach

**Does this PR introduce a user-facing change?**:

```release-note
new: introduce is_exe_upper_layer
```

### CODE REVIEW

**Comment by FedeDP** [comment]
This is really cool @loresuso ! Thanks for this effort!

**Review by FedeDP — COMMENTED** [review]
Changes LGTM; my only concern (and the same goes for exe_writable for example) is that we are returning "false" by default where unsupported (kernel versions < 3.18 using kmod), while we do not have any info really.
I don't think it is a major issue, but i don't really like returning wrong informations to clients.

File: userspace/libsinsp/threadinfo.cpp
Comment (loresuso) [reply userspace/libsinsp/threadinfo.cpp]: Of course we can! Updated!

**Comment by loresuso** [comment]
Hey @FedeDP, thank you! 
I agree with you, maybe we can better document the fact those new fields (`is_exe_writable` and `is_exe_upper_layer`) can be trusted only on specific kernel versions? I thought about that too, didn't know how to solve it honestly

**Comment by FedeDP** [comment]
> I agree with you, maybe we can better document the fact those new fields (is_exe_writable and is_exe_upper_layer) can be trusted only on specific kernel versions? I thought about that too, didn't know how to solve it honestly

Yeah that seems a reasonable thing to do.  
Another way of solving would be to use a string instead of a boolean and returning <NA> where the info is not available, but i don't quite like the idea :)

**Comment by FedeDP** [comment]
> Yeah, I don't like that too, execve gets called pretty much of course, let's not waste space on the ring buffer just for this :)

I meant that we can push onto the ring buffer eg: [ -1, 0, 1] where -1 means "NA", 0 means "false" and 1 means "true". Ring buffer would be untouched.
What i don't really like is having the filterchecks on a string instead of a boolean, it is much less elegant.

**Comment by loresuso** [comment]
Ah I got it now, sorry! Yeah you are right

File: driver/bpf/fillers.h
Comment (FedeDP) [inline driver/bpf/fillers.h:2099]: I see that basically until this loc, this functions shares 100% of the code with `get_exe_writable`.
I am not sure; is it costly to call all of these?
Can't we have a single function to extract both booleans?

**Comment by loresuso** [comment]
Hey @FedeDP, I think I have addressed all your comments:
- Stated that `is_exe_upper_layer` is meaningful if underlying kernel version is above 3.18
- Added a small function `get_exe_inode` to retrieve the inode of the executable. It is now used by `exe_writable` and `exe_upper_layer`. 
-  decided to add a function to properly set the `upper_layer` flag when scanning `procfs`. I tested it with docker and containerd and it seems perfectly ok, let me know what do you think on this
- Noted that  the overlays magic constant was defined in a c file before kernel 4.5, so to make it work, I defined PPM_OVERLAYFS_SUPER_MAGIC as you can see

Thank you! Your feedback is always much appreciated!!

File: userspace/libscap/scap_procs.c
Comment (LucaGuerra) [inline userspace/libscap/scap_procs.c:737]: Seems like some cases like this do not `fclose()` the file

File: userspace/libscap/scap_procs.c
Comment (LucaGuerra) [inline userspace/libscap/scap_procs.c:773]: If I understand the description right, that would be the part that fails in case Falco runs in a container. I believe it would need the right host prefix to work in that case if the appropriate directory is mounted. The other workaround would be to use `/proc/1/root/...` but I agree it's not the cleanest way of doing that.

File: userspace/libscap/scap_procs.c
Comment (loresuso) [reply userspace/libscap/scap_procs.c:743]: Thanks for noticing this!

File: userspace/libscap/scap_procs.c
Comment (loresuso) [reply userspace/libscap/scap_procs.c:776]: I found out the function `scap_get_host_root` and I think it solves the problem. Falco uses the environment variable `HOST_ROOT`, which is set to `/host` in the official docker image. I believe that if we mount directories such as `/var/lib/docker`, their path will follow this pattern too, ending up in `/host/var/lib/docker`. So just prepending it to each path should work, thank you! :)

Comment (loresuso) [reply userspace/libscap/scap_procs.c:776]: I am going to try it!

Comment (loresuso) [reply userspace/libscap/scap_procs.c:776]: I have tested it, and it now works also when Falco runs in a container if `/var/lib/docker` is mounted. Thank you!

**Comment by loresuso** [comment]
Hey maintainers! I have rebased this PR again, and I still think is a good feature to add. What do you think? 🙂

File: driver/bpf/fillers.h
Comment (FedeDP) [inline driver/bpf/fillers.h:2273]: I'd pre-initialize both to NULL so that even if `bpf_probe_read` fails for some reason, we still have a coherent behavior.

Moreover, i'm quite wary about adding the pointer arithmetic above :/ You know, that weird sense of fear with eBPF :D

Comment (FedeDP) [reply driver/bpf/fillers.h:2273]: I'd also double check the minimum kernel requirements to see if that holds true on any kernel supported (ie: >= 4.14).

File: driver/bpf/fillers.h
Comment (FedeDP) [inline driver/bpf/fillers.h:2261]: What about checking if `sb != NULL` here?
Same in the above usage inside `get_exe_writable` :)

File: driver/bpf/fillers.h
Comment (FedeDP) [inline driver/bpf/fillers.h:6301]: In `execve_family_flags` filler we actually check that `inode != NULL` before using it. Shall we do the same here?

File: driver/ppm_fillers.c
Comment (FedeDP) [inline driver/ppm_fillers.c:1279]: As for the eBPF probe, what about checking if `sb != NULL` ?

File: userspace/libscap/scap_procs.c
Comment (FedeDP) [inline userspace/libscap/scap_procs.c:695]: `fclose(mounts_file)` missing!

File: driver/bpf/fillers.h
Comment (FedeDP) [reply driver/bpf/fillers.h:6301]: Moreover i think you miss the `exe_upper_layer` in the `sched_prog_exec_4` filler (that is the new filler for exec for arm64 and s390x).

File: driver/ppm_fillers.c
Comment (FedeDP) [inline driver/ppm_fillers.c:1306]: This is ok.
Like for the eBPF probe, i think you miss the `exe_upper_layer` in the `f_sched_prog_exec` filler (that is the new filler for exec for arm64 and s390x).

**Review by FedeDP — CHANGES_REQUESTED** [review]

File: driver/bpf/fillers.h
Comment (incertum) [inline driver/bpf/fillers.h:2272]: Curiosity question: This double check condition is not redundant right? Asking because we don't seem to care about the actual fields and rather only use them to set the boolean value if there is an `upper_dentry`. Is it there is only ever either or?

 `bpf_fd_to_socktuple` seems to do similar pointer arithmetic.

**Review by incertum — COMMENTED** [review]
Thank you @loresuso for this :rocket: PR! Left only a few nit comments.
In addition, compiled and tested the actual new capability. Is working as expected :smile:, nice job!

**Comment by incertum** [comment]
One more thought/question: You mentioned that this does not work for all runtimes, would it make sense to mention the runtimes that are compatible in the definitions?

**Comment by incertum** [comment]
@loresuso as `modern_bpf` is one the near-term roadmap for initial experimental release, would it be possible to already add this new signal to `modern_bpf` as well? Would be much appreciated. CC @Andreagit97

CC @LucaGuerra would also appreciate adding the exe writable signal to `modern_bpf` (if possible) to have feature parity across all 3 drivers.

File: driver/bpf/fillers.h
Comment (Andreagit97) [reply driver/bpf/fillers.h:2159]: Agree with @incertum here we need to choose a way to handle this `NULL` check. To be honest, I would avoid the check unless it is necessary or explicitly required by the verifier... every check has a ridiculous cost but if we put many of them in the same flow it could impact perf.

File: driver/bpf/fillers.h
Comment (Andreagit97) [reply driver/bpf/fillers.h:2168]: `NULL` should be checked outside otherwise the verifier could go crazy with some clang/kernel versions, so yes if the verifier allows us to do it, we can return like this `return _READ(exe_file->f_inode);`

File: driver/bpf/fillers.h
Comment (Andreagit97) [reply driver/bpf/fillers.h:2159]: and adding them we increase also the bytecode complexity with many jumps

Comment (Andreagit97) [reply driver/bpf/fillers.h:2159]: Until we use the pointer only to perform another `_READ` we should be safe without the `NULL` check, this is the same logic I've followed in the modern probe, obviously here we are talking about lower kernel versions so never say never :roll_eyes:

File: driver/bpf/fillers.h
Comment (Andreagit97) [reply driver/bpf/fillers.h:2171]: yeah agree also on that we have just to understand in which order we want to merge them

File: driver/bpf/fillers.h
Comment (Andreagit97) [inline driver/bpf/fillers.h:2270]: Pain point here :/ the `struct ovl_inode` structure is already changed:
-> 4.14 https://elixir.bootlin.com/linux/v4.14/source/fs/overlayfs/ovl_entry.h#L60
-> 6.0.9 https://elixir.bootlin.com/linux/latest/source/fs/overlayfs/ovl_entry.h#L127

`struct inode *lower_inode` doesn't exist anymore :/

Comment (Andreagit97) [reply driver/bpf/fillers.h:2270]: it changes in kernel version `5.19`

Comment (incertum) [reply driver/bpf/fillers.h:2270]: See previous remark https://github.com/falcosecurity/libs/pull/287#discussion_r973520058

Let's just get rid of the redundant check :) tested in on a 6.0.8 kernel, works:

```
unsigned long sb_magic = _READ(sb->s_magic);
if(sb_magic == PPM_OVERLAYFS_SUPER_MAGIC)
{
	struct dentry *upper_dentry = NULL;
	char *vfs_inode = (char *)inode;
	bpf_probe_read(&upper_dentry, sizeof(upper_dentry), vfs_inode + sizeof(struct inode));
	if(upper_dentry)
	{
		return true;
	}
}
```

File: driver/bpf/fillers.h
Comment (incertum) [reply driver/bpf/fillers.h:2171]: We can merge this first if ok with @loresuso since it has been open for a while, happy to rebase the other PR then.

File: driver/bpf/fillers.h
Comment (Andreagit97) [reply driver/bpf/fillers.h:2270]: it would be amazing!!

Comment (loresuso) [reply driver/bpf/fillers.h]: Yes, it seems reasonable! Great Melissa, I think this is the best way to check this.
Updated with this change

**Comment by loresuso** [comment]
Hello everybody, I have reworked this PR addressing the comments and implemented the related version in the modern eBPF probe, so that we have feature parity in all three instrumentation modes.
I also tested the old eBPF probe and the kmod on older kernels, like version 4.18 and unfortunately we are not able to retrieve the upper layer information. It seems that the kernel uses a `sb_magic` that is different in newer kernel and associates it to ext3 in my experiments.  

If everyone like the approach of this PR, I can work on finding the minimum version from which it starts to be working and add a comment about it inside the code, but first I would like to hear everyone's opinion since this is a really time consuming task.

File: driver/bpf/fillers.h
Comment (Andreagit97) [inline driver/bpf/fillers.h:6459]:
```suggestion
		exe_writable = get_exe_writable(inode, cred);
```

**Comment by Andreagit97** [comment]
This PR LGTM! Thank you for all the effort you put into it! :tada: 

> If everyone like the approach of this PR, I can work on finding the minimum version from which it starts to be working and add a comment about it inside the code, but first I would like to hear everyone's opinion since this is a really time consuming task.

Yeah, it would be amazing to find the first working version!

> Also, I am thinking about ways to introduce tests for this feature. A possible way would be to mount an overlayfs and try to execute something from there and verify that the flag gets set.

Absolutely, I think we need a test, in this way we can easily detect when this access `vfs_inode + sizeof(struct inode));` is no more valid

File: driver/bpf/fillers.h
Comment (loresuso) [reply driver/bpf/fillers.h]: Thanks a lot!

**Review by Andreagit97 — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**
